nexus/networking/src/firewall_rules.rs

```diff
@@ -439,3 +486,100 @@ pub async fn plumb_service_firewall_rules(
+    nics: &[NetworkInterface],
+    allowed_ips: Vec<IpNet>,
+) -> Option<Vec<sled_agent_client::types::VpcFirewallRule>> {
+    if nics.is_empty() {
+        return None;
+    }
+
+    // Convert to host identifiers, how the sled-agent types expose these.
+    assert!(!allowed_ips.is_empty(), "Allowlist must not be empty here");
+    let filter_hosts = Some(
+        allowed_ips
+            .into_iter()
+            .map(|net| sled_agent_client::types::HostIdentifier::Ip(net.into()))
+            .collect(),
+    );
+
+    let low_priority_deny = sled_agent_client::types::VpcFirewallRule {
+        status: VpcFirewallRuleStatus::Enabled.into(),
+        direction: VpcFirewallRuleDirection::Inbound.into(),
+        targets: nics.to_vec(),
+        filter_hosts: filter_hosts.clone(),
+        filter_ports: Some(vec![L4PortRange {
+            first: 1.try_into().unwrap(),
+            last: u16::MAX.try_into().unwrap(),
+        }
+        .into()]),
+        filter_protocols: None,
+        action: VpcFirewallRuleAction::Deny.into(),
+        priority: ALLOW_LIST_LOW_PRIORTY,
+    };
+
+    let allow = sled_agent_client::types::VpcFirewallRule {
+        status: VpcFirewallRuleStatus::Enabled.into(),
+        direction: VpcFirewallRuleDirection::Inbound.into(),
+        targets: nics.to_vec(),
+        filter_hosts,
+
+        // TODO-correctness: Having no filter on ports or protocols is not
+        // techniically correct. Nexus should only receive TCP traffic, on HTTP
+        // or HTTPs traffic, for example. However, at this point in the system,
+        // we don't really have access to _which service_ this NIC is for, only
+        // that it is for _one_ of our services.
+        //
+        // We may want to reorganize this code to apply these rules on a
+        // per-service basis. This is discussed in RFD 474.
+        filter_ports: None,
+        filter_protocols: None,
+        action: VpcFirewallRuleAction::Allow.into(),
+        priority: ALLOWLIST_HIGH_PRIORITY,
+    };
+
+    Some(vec![low_priority_deny, allow])
+}
+
+/// Return the list of allowed IPs from the database.
+async fn lookup_allowed_source_ips(
+    datastore: &DataStore,
+    opctx: &OpContext,
+    log: &Logger,
+) -> Result<AllowedSourceIps, Error> {
+    match datastore.allowed_source_ips_view(opctx).await {
+        Ok(allowed) => {
+            debug!(log, "fetched allowlist from DB"; "allowed" => ?allowed);
+            let ips = if let Some(list) = allowed.allowed_ips {
+                AllowedSourceIps::List(
+                    list.into_iter().map(Into::into).collect(),
+                )
+            } else {
+                AllowedSourceIps::Any
+            };
+            Ok(ips)
+        }
+        Err(e) => {
+            error!(log, "failed to fetch allowlist from DB"; "err" => ?e);
+            Err(e)
+        }
+    }
+}
+
+/// FW rule priority for allow-list default-deny rule.
+const ALLOW_LIST_LOW_PRIORTY: u16 = u16::MAX - 1;
```

Comment:
How does these priorities interact with the "allow inbound HTTPS/HTTP/DNS from anywhere" rules that services also get?

Reply:
John and I talked offline about this, and he came up with a much better approach than the current one. Instead of adding any new rules at all, which seems like it would be more permissive in terms of ports and protocols than we want, let's _modify_ the existing rules. Those are the built-in Nexus and external DNS rules, part of the fixed data we insert into the DB initially.

To do that, we can look for _any_ rules that apply to the internal services VPC, and, if it allows any IP, change that to allow only those on the allowlist. We do not need a new default-deny rule, since OPTE has one at the bottom anyway.

We still need to worry about the case where we get an allow rule that has _some_ host filters. That's unexpected, since none do today, and we opted to just return an error in that case. It should help us catch changes to these VPC firewall rules that we don't expect right now.

Reply:
I implemented this in 381b1d2c0, where we add the host filters defined by the allowlist for the internal services VPC. It does the check for a _non-empty_ set of host filters, which we don't expect and don't overwrite, returning an error instead with some helpful breadcrumbs for future us.

Reply:
Awesome, thanks for the help with that design, it's much nicer than previously!